cmd/gardenlet/app/gardenlet.go

```diff
@@ -0,0 +1,458 @@
+		LeaderElection:         leaderElectionConfig,
+	}, nil
+}
+
+func (g *Gardenlet) cleanup() {
+	if err := os.RemoveAll(configv1alpha1.DefaultDiscoveryDir); err != nil {
+		g.Logger.Errorf("Could not cleanup base discovery cache directory: %v", err)
+	}
+}
+
+// Run runs the Gardenlet. This should never exit.
+func (g *Gardenlet) Run(ctx context.Context) error {
+	controllerCtx, controllerCancel := context.WithCancel(ctx)
+
+	defer controllerCancel()
+	defer g.cleanup()
+
+	leaderElectionCtx, leaderElectionCancel := context.WithCancel(context.Background())
+
+	// Prepare a reusable run function.
+	run := func(ctx context.Context) {
+		g.startControllers(ctx)
+	}
+
+	// If leader election is enabled, run via LeaderElector until done and exit.
+	if g.LeaderElection != nil {
+		g.LeaderElection.Callbacks = leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(_ context.Context) {
+				g.Logger.Info("Acquired leadership, starting controllers.")
+				run(controllerCtx)
+				leaderElectionCancel()
+			},
+			OnStoppedLeading: func() {
+				g.Logger.Info("Lost leadership, terminating.")
+				controllerCancel()
+			},
+		}
+		leaderElector, err := leaderelection.NewLeaderElector(*g.LeaderElection)
+		if err != nil {
+			return fmt.Errorf("couldn't create leader elector: %v", err)
+		}
+		leaderElector.Run(leaderElectionCtx)
+		return nil
+	}
+
+	// Leader election is disabled, thus run directly until done.
+	leaderElectionCancel()
+	run(controllerCtx)
+	return nil
+}
+
+func (g *Gardenlet) startControllers(ctx context.Context) {
+	controller.NewGardenletControllerFactory(
+		g.K8sGardenClient,
+		g.K8sGardenCoreInformers,
+		g.KubeInformerFactory,
+		g.Config,
+		g.Identity,
+		g.GardenNamespace,
+		g.Recorder,
+	).Run(ctx)
+}
+
+// We want to determine the Docker container id of the currently running Gardenlet because
+// we need to identify for still ongoing operations whether another Gardenlet instance is
+// still operating the respective Shoots. When running locally, we generate a random string because
+// there is no container id.
+func determineGardenletIdentity() (*gardencorev1alpha1.Gardener, string, error) {
+	var (
+		gardenletID     string
+		gardenletName   string
+		gardenNamespace = v1alpha1constants.GardenNamespace
+		err             error
+	)
+
+	gardenletName, err = os.Hostname()
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to get hostname: %v", err)
+	}
+
+	// If running inside a Kubernetes cluster (as container) we can read the container id from the proc file system.
+	// Otherwise generate a random string for the gardenletID
+	if cGroupFile, err := os.Open("/proc/self/cgroup"); err == nil {
+		defer cGroupFile.Close()
+		reader := bufio.NewReaderSize(cGroupFile, 4096)
+		line, con, err := reader.ReadLine()
+		if !con && err == nil {
+			splitBySlash := strings.Split(string(line), "/")
+			gardenletID = splitBySlash[len(splitBySlash)-1]
```

Comment:
The current reading of /proc/self/cgroup to find the container id (docker id) is fragile:
- The first line does not select a fixed cgroup hierarchy: on one cluster, I get “perf_event”; on another I get “devices”.
- Depending on the container runtime configuration, not all cgroup hierarchies are in a specific cgroup directory containing the container ID. It is safer to pick the line with “1:systemd“.

But even kubelet/docker/runc has two possible cgroup drivers: cgroupfs and systemd and they generate different path patterns. 

Example with cgroup driver = cgroupfs (`kubelet --cgroup-driver=cgroupfs` and `DOCKER_OPTS="--exec-opt native.cgroupdriver=cgroupfs"`):
```
1:name=systemd:/kubepods/burstable/pod533bebda-632d-4a45-9f32-237bcae5b1fc/f2c9e02d3140d5d72f23640b79596dcb043cc3ee818fb61507ff2dfb63dd0211
```
Example with cgroup driver = systemd (`kubelet --cgroup-driver=systemd` and `DOCKER_OPTS="--exec-opt native.cgroupdriver=systemd"`):
```
1:name=systemd:/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-pod1aecc643_23ea_11e9_beec_06c846f19394.slice/docker-da9977a4f9abe14ab2fa87d3780d92fd615b97cc3107fcd4a851f01857cb8ff8.scope
```

So I would recommend to use:
- `strings.TrimSuffix(gardenletID, ".scope")`
- `strings.TrimPrefix(gardenletID, "docker-")`
to handle both cases

Also, when switching to cgroup-v2 (https://github.com/kubernetes/enhancements/pull/1370 and https://github.com/kubernetes/kubernetes/pull/85218 ), it might be a single line starting with `0::`:
```
0::/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-pod1aecc643_23ea_11e9_beec_06c846f19394.slice/docker-da9977a4f9abe14ab2fa87d3780d92fd615b97cc3107fcd4a851f01857cb8ff8.scope
```

So I would recommend to pick the line starting with `0::` and if the gardenletID cannot be found (`len(gardenletID) != 64` or regexp `([0-9a-f]{64})`), fall back on the line starting with `1:name=systemd:`.

If it helps as a starting point, I used those regexp in another project:

```
cgroupRegexp1, _ = regexp.Compile("\n1:name=systemd:.*/kubepods.*[/-]pod([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}).*/([0-9a-f]{64})")
cgroupRegexp2, _ = regexp.Compile("\n1:name=systemd:.*/kubepods.*[/-]pod([a-f0-9]{8}_[a-f0-9]{4}_[a-f0-9]{4}_[a-f0-9]{4}_[a-f0-9]{12}).*/docker-([0-9a-f]{64})")
```

Reply:
Thanks @alban, will take a look.